torchsparse/backend/convolution/convolution_cuda.cu

```diff
@@ -52,8 +268,410 @@ __global__ void scatter_kernel(const int n_in, const int n_out, const int c,
+          -1);
+      kernel = torch::cat(
+          {kernel, torch::zeros({kernel.size(0), 8 - (kernel.size(1) % 8),
+                                 kernel.size(2)},
+                                options)},
+          1);
+    }
+    if (out_feat.size(1) % 8 != 0) {
+      out_feat = torch::cat(
+          {out_feat, torch::zeros({out_feat.size(0), 8 - (in_feat.size(1) % 8)},
+                                  options)},
+          -1);
+      kernel = torch::cat({kernel, torch::zeros({kernel.size(0), kernel.size(1),
+                                                 8 - (kernel.size(2) % 8)},
+                                                options)},
+                          -1);
+    }
+  } else {
+    if (in_feat.size(1) % 4 != 0) {
+      in_feat = torch::cat(
+          {in_feat,
+           torch::zeros({in_feat.size(0), 4 - (in_feat.size(1) % 4)}, options)},
+          -1);
+      kernel = torch::cat(
+          {kernel, torch::zeros({kernel.size(0), 4 - (kernel.size(1) % 4),
+                                 kernel.size(2)},
+                                options)},
+          1);
+    }
+    if (out_feat.size(1) % 4 != 0) {
+      out_feat = torch::cat(
+          {out_feat, torch::zeros({out_feat.size(0), 4 - (in_feat.size(1) % 4)},
+                                  options)},
+          -1);
+      kernel = torch::cat({kernel, torch::zeros({kernel.size(0), kernel.size(1),
+                                                 4 - (kernel.size(2) % 4)},
+                                                options)},
+                          -1);
+    }
+  }
+  // std::cout << in_feat.size(0) << " " << in_feat.size(1) << " " <<
+  // out_feat.size(0) << " " << out_feat.size(1) << " " << kernel.size(0) << " "
+  // << kernel.size(1) << " " << kernel.size(2) << std::endl;
+
+  int n_in_feats = in_feat.size(0);
+  int n_in_channels = in_feat.size(1);
+  int n_out_feats = out_feat.size(0);
+  int n_out_channels = out_feat.size(1);
+
+  int kernel_volume = kernel.size(0);
+
+  // memory optimization
+  bool precompute_mid = false;
+  // possibly in the last position
+  int mid_kernel = conv_mode == 2 ? kernel_volume - 1 : kernel_volume / 2;
+  int max_kmap_size = 1;
+  // we can precompute features for w[0,0] which avoids gather/scatter
+  if (kernel_volume % 2 == 1 && n_in_feats == n_out_feats) {
+    precompute_mid = true;
+    max_kmap_size =
+        *std::max_element(neighbor_offset.data_ptr<int>(),
+                          neighbor_offset.data_ptr<int>() + kernel_volume / 2);
+    max_kmap_size =
+        std::max(max_kmap_size,
+                 *std::max_element(
+                     neighbor_offset.data_ptr<int>() + kernel_volume / 2 + 1,
+                     neighbor_offset.data_ptr<int>() + kernel_volume));
+    max_kmap_size = std::max(max_kmap_size, 1);
+    // (N, c) X (c, o) = (N, o)
+    // torch::mm_out(out_feat, in_feat, kernel[mid_kernel]);
```

Comment:
We can remove line 500-501 to avoid possible confusion. Currently we are doing the computation for kernel offset = (0, 0, 0) in the last, instead of in the first.